[MDS-6086] Updated permit condition extraction to happen asynchronously using Celery

simensma-fresh opened the pull request:
## Objective 

[MDS-6086](https://bcmines.atlassian.net/browse/MDS-6086)

- Added a celery worker for local development that handles permit condition extraction jobs
- Updated permit condition extraction endpoint to create a celery job
- Added endpoints to retrieve the status and results of a permit condition job
- A couple of tweaks to the test reporting scripts to fix a couple of issues that popped up when running it against large permits

![image](https://github.com/user-attachments/assets/3115332c-049f-47ab-a346-e7b261e5c949)

simensma-fresh committed:
[MDS-6086] Added support for async permit condition extraction using celery

simensma-fresh committed:
[MDS-6086] Updated documentation of permit condition extraction endpoints

simensma-fresh committed:
Cleanup

taraepp commented on services/permits/app/compare_extraction_results.py:
```diff
@@ -193,6 +214,8 @@ def validate_condition(csv_pairs):
             key=lambda c: c.get("Key"),
         )
 
+        print([c.get("Key") for c in match_results["context_comparison_results"]])
```
this log looks more like it was left in by mistake?

simensma-fresh replied:
Definitely, removed it.

taraepp commented on services/permits/app/permit_conditions/converters/pdf_to_text_converter.py:
```diff
@@ -62,9 +66,9 @@ def _read_pdf(self, file_path: Path) -> List[str]:
-                            extraction_mode="layout",
-                            layout_mode_space_vertically=False,
-                            layout_mode_scale_weight=0.5,
+                            # extraction_mode="",
```
?

taraepp commented on services/permits/app/permit_conditions/resources/permit_condition_resource.py:
```diff
@@ -42,56 +50,83 @@ async def extract_permit_conditions(file: UploadFile = File(...)) -> PermitCondi
+    if res.status == "SUCCESS":
+        return res.get()
+    else:
+        raise HTTPException(400, detail=f"Task has not completed successfully. Current status: {res.status}")
```
I think I'd expect to get a 400 if I sent a bad task id, but I wonder if there's something more appropriate to send back to indicate "still working on it". Maybe `202 Accepted`? Or a 500 if it's a bad status?

simensma-fresh replied:
Good call, 202 seems good for the "processing" stage of it.
I updated it to be:
`200` -> Task finished
`202` -> Task created, but not finished
`500` -> Task failed

simensma-fresh committed:
MDS-6086 Tweaks after PR feedback

taraepp commented on services/permits/app/permit_conditions/resources/permit_condition_resource.py:
```diff
@@ -42,56 +50,86 @@ async def extract_permit_conditions(file: UploadFile = File(...)) -> PermitCondi
+    elif res.status == "FAILURE":
+        raise HTTPException(500, detail=f"Task failed to complete: {res.status}")
+    else:
+        raise HTTPException(202, detail=f"Task has not completed yet. Current status: {res.status}")
```
Does it work to return a 200-type code with HTTPException? If it works, cool, but it looks kinda fishy to me. 😄 

May want to consider an enum with SUCCESS/FAILURE/PROGRESS statuses.

simensma-fresh replied:
It does indeed work, but yup not the most clear.

asinn134 commented on services/permits/app/compare_extraction_results.py:
```diff
@@ -103,6 +117,8 @@ def validate_condition(csv_pairs):
         auto_extracted_content = create_content_instances(auto_extracted_df)
         manual_extracted_content = create_content_instances(manual_extracted_df)
 
+        print(f'Found {len(auto_extracted_content)} conditions in {auto_extracted_csv} and {len(manual_extracted_content)} conditions in {manual_extracted_csv}')
```
Was this print statement meant to be added as well?

simensma-fresh replied:
Indeed it was intentional

simensma-fresh committed:
MDS-6086 Added permit condition resource tests, improved return values + add enum

simensma-fresh commented:
@taraepp I updated the response models (the 202 now shows up in the swagger docs as well) + added an enum. Also added some missing tests.

![image](https://github.com/user-attachments/assets/7e66a6e4-8506-4848-a585-d969f8aad784)
